File: app/src/main/java/com/maciejkozlowski/databases/results/Saver.java
package com.maciejkozlowski.databases.results;

import android.os.Environment;
import android.util.Log;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.List;


/**
 * Created by Maciej on 2017-05-10.
 */

public class Saver {

    public static final String FILE_NAME = "results.csv";
    public static final String TAG = "###";
    public static final String SEPARATOR = ";";
    public static final String NEW_LINE = "\n";

    private FileOutputStream outputStream;
    private boolean opened;

    public void open() {
        if (!opened) {
            try {
                File file = new File(Environment.getExternalStorageDirectory(), FILE_NAME);
                if (file.exists()) {
                    file.delete();
                }

                outputStream = new FileOutputStream(file);
                opened = true;
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    public void write(String data) {
        if (opened) {
            try {
                outputStream.write(data.getBytes());
            } catch (Exception ex) {
                ex.printStackTrace();
            }
        } else {
            Log.e(TAG, "NOT OPENED");
        }
    }

    public void close() {
        if (opened) {
            try {
                outputStream.close();
                opened = false;
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    public void save(ResultSet resultSet) {
        open();

        saveResult(resultSet.getCreating());
        saveResult(resultSet.getReading());
        saveResult(resultSet.getDeleting());
        saveResult(resultSet.getUpdating());

        close();
    }

    private void saveResult(Result result) {
        write(result.getType() + NEW_LINE);
        for (int size : Test.SIZES) {
            for (String type : Test.TYPES) {
                String key = size + "-" + type;
                List<Long> times = result.getTimes().get(key);
                write(key + SEPARATOR);
                for (long time : times) {
                    write(time + SEPARATOR);
                }
            }
            write(NEW_LINE);
        }
        write(NEW_LINE);
        write(NEW_LINE);
    }
}

File: app/src/main/java/com/maciejkozlowski/databases/results/Saver.kt
package com.maciejkozlowski.databases.results

import android.os.Environment
import android.util.Log

import java.io.File
import java.io.FileOutputStream
import java.io.IOException


/**
 * Created by Maciej on 2017-05-10.
 */

class Saver {

    private var outputStream: FileOutputStream? = null
    private var opened: Boolean = false

    fun open() {
        if (!opened) {
            try {
                val file = File(Environment.getExternalStorageDirectory(), FILE_NAME)
                if (file.exists()) {
                    file.delete()
                }

                outputStream = FileOutputStream(file)
                opened = true
            } catch (e: IOException) {
                e.printStackTrace()
            }

        }
    }

    fun write(data: String) {
        if (opened) {
            try {
                outputStream!!.write(data.toByteArray())
            } catch (ex: Exception) {
                ex.printStackTrace()
            }

        } else {
            Log.e(TAG, "NOT OPENED")
        }
    }

    fun close() {
        if (opened) {
            try {
                outputStream!!.close()
                opened = false
            } catch (e: IOException) {
                e.printStackTrace()
            }

        }
    }

    fun save(resultSet: ResultSet) {
        open()

        saveResult(resultSet.creating)
        saveResult(resultSet.reading)
        saveResult(resultSet.deleting)
        saveResult(resultSet.updating)

        close()
    }

    private fun saveResult(result: Result) {
        write(result.type + NEW_LINE)
        for (size in TestConstants.SIZES) {
            for (type in TestConstants.TYPES) {
                val key = size.toString() + "-" + type
                val times = result.times[key]
                write(key + SEPARATOR)
                for (time in times!!) {
                    write(time.toString() + SEPARATOR)
                }
            }
            write(NEW_LINE)
        }
        write(NEW_LINE)
        write(NEW_LINE)
    }

    companion object {
        const val FILE_NAME = "results.csv"
        const val TAG = "###"
        const val SEPARATOR = ";"
        const val NEW_LINE = "\n"
    }
}
